src/repr/src/row.rs

```diff
@@ -1193,6 +1328,161 @@ impl RowPacker<'_> {
+    ///   that it is possible to generate semantically empty ranges with this
+    ///   function, but they will _not_ be output as empty ranges.
+    ///
+    /// # Panics
+    /// - If any `Datum::Null` were pushed using `f`.
+    /// - If the number of `Datum`s pushed using `f` is incommensurate with the
+    ///   provided `RangeBoundDesc` or the structure of ranges in general.
+    /// - If `Datum`s of different types were pushing using `f`.
+    pub fn push_range_with<F, R>(
+        &mut self,
+        RangeBoundDesc {
+            inf: lower_inf,
+            inc: lower_inc,
+        }: RangeBoundDesc,
+        RangeBoundDesc {
+            inf: upper_inf,
+            inc: upper_inc,
+        }: RangeBoundDesc,
+        f: F,
+    ) -> Result<R, RangeError>
+    where
+        F: FnOnce(&mut RowPacker) -> R,
+    {
+        let start = self.row.data.len();
+        self.row.data.push(Tag::Range.into());
+
+        let mut flags: u8 = 0;
+        let mut expected_datums = 0;
+
+        for (val, set, expected_datum_diff) in [
+            (lower_inf, range::RANGE_LB_INF, 1),
+            (lower_inc, range::RANGE_LB_INC, 0),
+            (upper_inf, range::RANGE_UB_INF, 1),
+            (upper_inc, range::RANGE_UB_INC, 0),
+        ] {
+            if val {
+                flags |= set;
+            } else {
+                expected_datums += expected_datum_diff;
+            }
+        }
+
+        self.row.data.push(flags);
+        let mut datum_check = self.row.data.len();
+
+        let out = f(self);
+
+        // Validate that what was written maintains the correct invariants.
+        let mut actual_datums = 0;
+        let mut seen = None;
+        while datum_check < self.row.data.len() {
+            let d = unsafe { read_datum(&self.row.data, &mut datum_check) };
+            match seen {
+                None => seen = Some(d),
+                Some(seen) => {
+                    let seen_kind = DatumKind::from(seen);
+                    let d_kind = DatumKind::from(d);
+                    if seen_kind != d_kind {
+                        panic!("range contains inconsistent data; expected {seen_kind:?} but got {d_kind:?}");
+                    }
+
+                    if seen > d {
+                        self.row.data.truncate(start);
+                        return Err(RangeError::MisorderedRangeBounds);
+                    }
+                }
+            }
+            actual_datums += 1;
+        }
+
+        if actual_datums != expected_datums {
+            panic!(
+                "range contains the wrong number of datums; expected {expected_datums} but got {actual_datums}"
+            );
+        }
+
+        // Anything that triggers this check is undefined behavior, so
+        // unnecessary but also trivial to perform the check in our case.
+        if datum_check != self.row.data.len() {
+            panic!("non-Datum data packed into row");
+        }
+
+        Ok(out)
+    }
```

Comment:
These two APIs seem a little easy to misuse, in opposite ways. The first because `Datum::Null` means infinite, and the second because you can do lots of bad with the `RowPacker` you're handed. Is there an immediate reason we need to support `push_range_with`? We need it for lists and maps because they can be arbitrarily nested, but we're not planning to support range types over list, maps, or arrays, right?

The API that I'd suggest is

```rust
enum RangeBoundDesc {
    Finite {
        value: Datum<'static>,
        inclusive: bool,
    },
    Infinite,
}

pub fn push_range(
    &mut self,
    lower: RangeBoundDesc,
    upper: RangeBoundDesc,
) -> Result<(), RangeError>;
```

but you might have context about what's coming in the next PR that I don't!

Reply:
Oh, I see, the answer was in this PR, and it's `try_push_proto`!  In that case I'd suggest an API like:

```rust
enum RangeBoundDesc<D> {
    Finite {
        value: D,
        inclusive: bool,
    },
    Infinite,
}

pub fn push_range<'a>(
    &mut self,
    lower: RangeBoundDesc<Datum<'a>>,
    upper: RangeBoundDesc<Datum<'a>>,
) -> Result<(), RangeError>;

pub fn push_range_with<L, U, E>(
    &mut self,
    lower: RangeBoundDesc<L>,
    upper: RangeBoundDesc<U>,
) -> Result<(), E>
where
    L: FnOnce(&mut RowPacker) -> Result<(), E>,
    U: FnOnce(&mut RowPacker) -> Result<(), E>,
    E: From<RangeError>;
```

Reply:
Yeah, this was a bit of a mess for want of supporting `Null` values as bounds; I removed the ability to use those but kept this structure for sake of structuring the input, while not making another mess if it turns out using `Datum::Null` is a hard requirement.